agent/service.go

```diff
@@ -342,27 +330,10 @@ func (ts *TaskService) Shutdown(ctx context.Context, req *shimapi.ShutdownReques
 		log.G(ctx).WithError(err).Warn("error cleaning up")
 	}
 
-	// TODO: cleaner method of shutdown
-	resp, err := ts.runc.Shutdown(ctx, req)
-	if err != nil {
-		log.G(ctx).WithError(err).Error("shutdown failed")
-		return nil, err
-	}
-
-	log.G(ctx).Debug("shutdown succeeded")
-	return resp, nil
-}
-
-func (ts *TaskService) StartShim(ctx context.Context, id, containerdBinary, containerdAddress string) (string, error) {
-	log.G(ctx).WithFields(logrus.Fields{"id": id, "bin": containerdAddress, "addr": containerdAddress}).Debug("start_shim")
-
-	ctx = namespaces.WithNamespace(ctx, defaultNamespace)
-	resp, err := ts.runc.StartShim(ctx, id, containerdBinary, containerdAddress)
-	if err != nil {
-		log.G(ctx).WithError(err).Error("start shim failed")
-		return "", err
-	}
+	// We don't want to call runc.Shutdown here as it just os.Exits behind.
+	// Invoking cancel here for gracefull shutdown instead and call runc Shutdown at end.
+	ts.cancel()
 
-	log.G(ctx).Debugf("start shim succeeded: %s", resp)
-	return resp, err
+	log.G(ctx).Debug("going to gracefully shutdown agent")
+	return &types.Empty{}, nil
```

Comment:
Out of curiosity, what is the `&types.Empty{}` usage for? Is it ever to be used?

Reply:
No, it indicates an empty response.
See here https://github.com/gogo/protobuf/blob/master/types/empty.pb.go#L37

Reply:
Ah, I see. Things are modeled as a request response operations. In this case, shutdown doesn't have a modeled response so an `Empty` one is used. Thanks! :)